Automattic/simplenote-android
# Add Sync Persistence

### Fix
Add persistence to the note last sync time cache added in https://github.com/Automattic/simplenote-android/pull/1140.  Since that pull request has not been merged, I'm targeting the associated branch, `add/last-sync-time-indicator`.  These changes save the last sync time for each note when a remote change occurs with the `onNetworkChange` callback and when a local change has been confirmed as received by the server with the `onSyncObject` callback.  The last sync times are saved as preferences.  The preferences option was chosen for storage since preferences are meant to store private, primitive data in key-value pairs and all that is needed for the last sync time data is a note ID as a `String` (i.e. the key) and a time as a `long` (i.e. the value).  Originally, a [Room database](https://developer.android.com/training/data-storage/room) was used for storage.  The storage was changed to preferences since databases are meant for structured data, only one item will be retrieved at a time using the key, the note ID and last sync time are not sensitive data, and using Room would have required the [`allowMainThreadQueries()` method](https://developer.android.com/reference/androidx/room/RoomDatabase.Builder#allowMainThreadQueries()) for executing queries in order to avoid using an [`AsyncTask`, which has been deprecated](https://developer.android.com/reference/android/os/AsyncTask).

#### Note
For some unknown reason, all notes do not trigger the `onNetworkChange` callback when first loading notes after login.  When testing various accounts, approximately 70% of notes triggered `onNetworkChange` and caused a last sync time to be saved for those notes.  All other notes that did not trigger `onNetworkChange` would not have a last sync time until those notes were edited.  Similarly, if a user is already logged in when installing the app with these changes, notes will not have a last sync time value cached or saved until the note is edited remotely or locally and the device has an active network connection.  Since the last sync time is most important after an edit has been made and these changes will update the last sync time after an edit, solving the problems with first loading notes after login and determining a last sync time for users already logged in can be addressed in a future enhancement.

### Test
Since the last sync time cache can be affected by local and remote edits, using two devices to test is best.  I used an Android emulator and the web app.
1. **Android**: Tap any note in list.
2. **Web**: Tap note in list from Step 1.
3. **Android**: Tap ***Information*** action in app bar.
4. **Android**: Notice ***Synced*** field is hidden.
5. **Android**: Tap anywhere outside ***Information*** sheet to dismiss.
6. **Web**: Edit note from Step 1.
7. Wait ten seconds.
8. **Android**: Tap ***Information*** action in app bar.
9. **Android**: Notice ***Synced*** field is shown.
10. **Android**: Enable airplane mode.
11. **Android**: Edit note from Step 1.
12. **Android**: Tap ***Information*** action in app bar.
13. **Android**: Notice ***Synced*** field is shown and unchanged.
14. **Android**: Notice ***Modified*** field is shown and changed.
15. **Android**: Tap anywhere outside ***Information*** sheet to dismiss.
16. Wait one minute.
17. **Android**: Disable airplane mode.
18. Wait ten seconds.
19. **Android**: Tap ***Information*** action in app bar.
20. **Android**: Notice ***Synced*** field is shown and changed.
21. **Android**: Notice ***Modified*** field is shown and unchanged.
22. **Android**: Tap recents/overview system navigation button.
23. **Android**: Swipe app away from recents/overview list.
24. **Android**: Launch app.
25. **Android**: Tap note from Step 1.
26. **Android**: Tap ***Information*** action in app bar.
27. **Android**: Notice ***Synced*** field is shown with value from Step 20.

### Review
Only one developer is required to review these changes, but anyone can perform the review.  @dmsnell, I requested you as a reviewer since you created the last sync time cache.  Don't feel obligated to review this pull request, but you can if you want.  I just wanted to notify you of this enhancement to your last sync time cache feature.

### Release
These changes do not require release notes.
